Volatile is not the same as no cache (#120)
UPB-FILS-MA/upb-fils-ma.github.io: slides/lectures/02/mmio/slides.md

|  |  |
|----------|-------------|
| `read_volatile`, `write_volatile` | **no** compiler **optimization** |
| `read`, `write`, `*p` | **use** compiler **optimization**  |

::right::

![SysCtrl Registers](/mmio/sysctrl_registers.png)

---
---
# 8 bit processor
with cache

![8 Bit Processor](/mmio/8-bit-processor-with-cache.svg)

---
---
# No Cache or Flush Cache

- Cache types:

>   - *write-trough* - data is written to the cache and to the main memory (bus)

Reviewer: ```suggestion
  - *write-through* - data is written to the cache and to the main memory (bus)
```

  - *write-back* - data is written to the cache and later to the main memory (bus)
- few Cortex-M MCUs have cache

> - the Memory Mapped region is set as *nocache*

Reviewer: ```suggestion
- the Memory Mapped I/O region is set as *nocache*
```

- for chips that use cache
  - *nocache* regions have to be set manually (if MCU knows)

>   - the cache has to be flushed before every `volatile_read` and after `volatile_write`

Reviewer: ```suggestion
  - or, the cache has to be flushed before a `volatile_read` and after a `volatile_write`
  - beware DMA controllers that can't see the cache contents
```
